File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int dfs(TreeNode node) {
        // 리프 노드에 도달하면 left와 right는 0을 반환한다.
        if(node == null) {
            return 0;
        }

        // 왼쪽과 오른쪽 자식의 높이를 반환한다.
        int left = dfs(node.left);
        int right = dfs(node.right);

        // 각 자식의 균형이 이미 깨져있거나, 높이차가 1보다 크다면 균형이 깨졌다는 의미로 -1을 반환한다.
        if(left == -1 || right == -1 || Math.abs(left - right) > 1) {
            return -1;
        }

        // 현재 노드의 높이를 반환한다.
        return Math.max(left, right) + 1;
    }

    public boolean isBalanced(TreeNode root) {
        return dfs(root) != -1;
    }
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        fun dfs(node: TreeNode?): Int {
            // 예외 처리
            if(node == null) {
                return 0
            }
            
            // 재귀 dfs로 왼쪽 리프 노드까지 탐색
            val left = dfs(node.left)
            
            // 재귀 dfs로 오른쪽 리프 노드까지 탐색
            val right = dfs(node.right)
            
            // 높이 균형이 아닌 경우 -1 반환
            return if(left == -1 || right == -1 || abs(left - right) > 1) 
                -1
            else 
                // 현재 노드의 높이 반환
                left.coerceAtLeast(right) + 1
        }
        
        // 반환 값이 -1이라면 균형이 깨졌으므로 false를 반환한다. 아니라면 true를 반환한다.
        return dfs(root) != -1
    }
}